### PR TITLE
Sim WString replace and fixes

### LINKED ISSUE #24: OffBy1 Error in WString implementation for HALSim
Operator overload for += ```String& operator+=(const String& str)``` calculates the new buffer size as follows:
```char* tmp = new char[m_size + str.m_size];```

The new size therefor includes space for 2 null_termination chars and is 1 byte to big.
Correct buffer size would be ```m_size + str.m_size - 1```

Same error persist for new m_size calculation!

### PR DESCRIPTION
**Pull request by LuDubies**
Closes #24

### CODE REVIEW

File: lib/ArduinoNative/src/WString.h [hunk 1]
```diff
@@ -478,6 +499,103 @@ class String
...
+            if (0 < accumulatedLengthDiff)
+            {
+                /* replace m_buffer with increased size buffer */
+                char* tmp = new char[m_size + accumulatedLengthDiff];
+                if (nullptr != tmp)
+                {
+                    char* writeTo = tmp;
+
+                    /* write new buffer from original and replacement */
+                    while ((foundAt = strstr(readFrom, find.m_buffer)) != NULL)
```
Comment (gabryelreyes) [inline lib/ArduinoNative/src/WString.h:581]: Same as before

**Review by gabryelreyes — CHANGES_REQUESTED** [review]

**Comment by gabryelreyes** [comment]
Can you please add a test for when the original string is empty?

File: lib/ArduinoNative/src/WString.h [hunk 2]
```diff
@@ -290,7 +291,27 @@ class String
...
+        }
+
+        return *this;
+    }
+
+    String& operator+=(const char* cstr)
+    {
+        if (nullptr != cstr)
+        {
+            char* tmp = new char[m_size + strlen(cstr)];
```
[the remark this answers is not in the document]
Comment (LuDubies) [reply lib/ArduinoNative/src/WString.h:305]: strlen() will return the lenght without the null terminator, so we dont count it double anyways.

File: test/test_WString/test_WString.cpp [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    tester.replace("a", "AAA");
+    TEST_ASSERT_EQUAL_STRING("I AAAm AAA long string thAAAt will feAAAture some replAAAcements!", tester.c_str());
+    TEST_ASSERT_EQUAL(original.length() + 10, tester.length());
+
+    /* replacement with empty string */
+    tester = String(original);
+    tester.replace(" ", "");
+    TEST_ASSERT_EQUAL_STRING("Iamalongstringthatwillfeaturesomereplacements!", tester.c_str());
+
+    return;
```
[the remark this answers is not in the document]
Comment (LuDubies) [reply test/test_WString/test_WString.cpp:143]: See test_Util

File: lib/ArduinoNative/src/WString.h [hunk 4]
```diff
@@ -290,7 +291,27 @@ class String
...
+
+            if (nullptr != tmp)
+            {
+                strcpy(tmp, m_buffer);
+                strcat(tmp, cstr);
+
+                delete[] m_buffer;
+                m_buffer = tmp;
+                m_size   = m_size + strlen(cstr);
```
Comment (nhjschulz) [inline lib/ArduinoNative/src/WString.h:314]: repetitive use of strlen(cstr) - performance !

Better 
size_t len = (nullptr != cstr) ? strlen(cstr) : 0U;

if  (0U != len) 
{ 
  .....

File: lib/ArduinoNative/src/WString.h [hunk 5]
```diff
@@ -290,7 +291,27 @@ class String
...
+    String& operator+=(const char* cstr)
+    {
+        if (nullptr != cstr)
+        {
+            char* tmp = new char[m_size + strlen(cstr)];
+
+            if (nullptr != tmp)
+            {
+                strcpy(tmp, m_buffer);
+                strcat(tmp, cstr);
```
Comment (nhjschulz) [inline lib/ArduinoNative/src/WString.h:310]: faster then strcat():
strcpy( & tmp[m_size], cstr);

**Review by nhjschulz — COMMENTED** [review]
Wouln't it be better to wrap  std::string with WString instead of reimplementing this the hard and ugly  char buffer way ?

File: lib/ArduinoNative/src/WString.h (hunk 5 above)
Comment (LuDubies) [reply lib/ArduinoNative/src/WString.h:310]: m_size index would be off by 1 because of the null termination!!
A working version would be ```strcpy(&tmp[m_size-1], cstr)```, but then the strcat code looks way cleaner!

Comment (nhjschulz) [reply lib/ArduinoNative/src/WString.h:310]: sure cleaner but ssssllloooowwwwwweeeeeeerrrrrrrr.  strcat() needs to scan to the end of the destination string and then do a strcpy() there. As we know the offset to copy already, we can  save the world some energy.

File: lib/HALTarget/src/Network.cpp [hunk 6]
```diff
@@ -95,7 +95,7 @@ bool Network::process()
         /* Nothing to do. */
         isSuccess = true;
         break;
-    
+
     case STATE_SETUP:
```
Comment (BlueAndi) [inline lib/HALTarget/src/Network.cpp:99]: Changes in this file are not related to the PR title.

File: test/test_WString/test_WString.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (BlueAndi) [reply test/test_WString/test_WString.cpp:150]: Remove "return" without value.

File: test/test_WString/test_WString.cpp [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+/**
+ * Test WString += implementation.
+ */
+static void testWStringAppend(void)
+{
+    String original = String("Im short");
+    char*  classic  = new char[4];
```
Comment (BlueAndi) [inline test/test_WString/test_WString.cpp:159]: Why using heap memory, which btw. is nevered freed?

File: lib/ArduinoNative/src/WString.h [hunk 8]
```diff
@@ -478,6 +499,103 @@ class String
...
+     *
+     * @param[in] find      Pattern
+     * @param[in] replace   Replacement for Pattern
+     */
+    void replace(const String& find, const String& replace)
+    {
+        unsigned int replaceLen = replace.length();
+        unsigned int findLen    = find.length();
+
+        if ((0 == length()) || (0 == findLen))
```
Comment (BlueAndi) [inline lib/ArduinoNative/src/WString.h:513]: m_buffer can be a nullptr

**Review by BlueAndi — CHANGES_REQUESTED** [review]

File: test/test_WString/test_WString.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (LuDubies) [reply test/test_WString/test_WString.cpp:150]: Done in [7131ee2](https://github.com/BlueAndi/DroidControlShip/pull/30/commits/7131ee2854693237c046fa6e71c81362bf947c92)

File: lib/ArduinoNative/src/WString.h (hunk 8 above)
Comment (LuDubies) [reply lib/ArduinoNative/src/WString.h]: Resolved with [0a851f2](https://github.com/BlueAndi/DroidControlShip/pull/30/commits/0a851f278e883ffcb55ac8a7f9862882d4d7a118)

File: lib/HALTarget/src/Network.cpp (hunk 6 above)
Comment (LuDubies) [reply lib/HALTarget/src/Network.cpp]: Resolved with [894deb3](https://github.com/BlueAndi/DroidControlShip/pull/30/commits/894deb35cdc54e80f856a7d02bf2ef8635dd5504)

**Review by BlueAndi — APPROVED** [review]

File: lib/ArduinoNative/src/WString.h (hunk 5 above)
Comment (LuDubies) [reply lib/ArduinoNative/src/WString.h]: Implemented in [bf66f40](https://github.com/BlueAndi/DroidControlShip/pull/30/commits/bf66f40a8fcd14f0ad79000dcc651935578d2192)

**Review by gabryelreyes — APPROVED** [review]
